### PR TITLE
DM-228 reporting metrics data model and resource

### PR DESCRIPTION
**Pull request by sutherlanda**
Implements a counter that can be used to track views on opportunities, etc.

### CODE REVIEW

File: src/back-end/lib/db/counter.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { tryDb } from 'back-end/lib/db';
+import { valid } from 'shared/lib/http';
+
+interface ViewCounter {
+  name: string;
+  count: number;
+}
```
Comment (dhruvio) [inline src/back-end/lib/db/counter.ts]: Rename to `Counter` and move to shared counter resource

File: src/back-end/lib/permissions.ts [hunk 2]
```diff
@@ -204,3 +204,9 @@ export async function editCWUProposal(connection: Connection, session: Session,
 export async function deleteCWUProposal(connection: Connection, session: Session, proposalId: string): Promise<boolean> {
   return session.user && await isCWUProposalAuthor(connection, session.user, proposalId) || false;
 }
+
+// Metrics.
+
+export function readAllCounters(session: Session): boolean {
+  return isAdmin(session);
+}
```
Comment (dhruvio) [inline src/back-end/lib/permissions.ts]: I may not have explained this in the ticket, so my apologies if that's the case. Can we add a separate function called `readManyCounters`, and allow all admins & gov users to do so? This permissions check should happen at the beginning of the `readMany` CRUD actions. Thinking is that non-admin CWU gov opp owners will need to see an opportunity's total views.

File: src/back-end/lib/db/counter.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  // Create new counters where applicable
+  for (const name of names) {
+    if (!existingCounters.includes(name)) {
+      await connection<ViewCounter>('viewCounters')
+        .insert({
+          name,
+          count: 0
+        });
+    }
+  }
```
Comment (dhruvio) [inline src/back-end/lib/db/counter.ts]: Doesn't look like this is used?

File: src/back-end/lib/db/counter.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  name: string;
+  count: number;
+}
+
+export const incrementCounters = tryDb<[string[]], Record<string, number>>(async (connection, names) => {
+  // Update existing counters
+  const existingCounters = await connection<ViewCounter>('viewCounters')
+    .whereIn('name', names)
+    .increment('count')
+    .update({}, 'name');
```
Comment (dhruvio) [inline src/back-end/lib/db/counter.ts]: Strange that knex requires you to call increment then update.

File: src/shared/lib/resources/counter.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { BodyWithErrors } from 'shared/lib/types';
+
+export interface UpdateValidationErrors extends BodyWithErrors {
+  name?: string[];
+}
```
Comment (dhruvio) [inline src/shared/lib/resources/counter.ts]: Can you add `export type UpdateRequestBody = null;` for completeness? Front-end will use it.

File: src/back-end/lib/db/counter.ts (hunk 3 above)
Comment (sutherlanda) [reply src/back-end/lib/db/counter.ts]: No, but the ticket specified it, so I assumed it was for future use.  Should I remove?

File: src/back-end/lib/db/counter.ts (hunk 4 above)
Comment (sutherlanda) [reply src/back-end/lib/db/counter.ts]: It doesn't require it, but increment only returns # of incremented rows.  I use update here to get the names of the updated rows so I know which counters needed to be inserted in the next step.

File: src/back-end/lib/permissions.ts (hunk 2 above)
Comment (sutherlanda) [reply src/back-end/lib/permissions.ts]: No problem, I will add this.  I'll take opportunityId from the name of the counter provided to determine whether a gov user is an owner of the opportunity corresponding to the counter.

Comment (sutherlanda) [reply src/back-end/lib/permissions.ts]: Actually, I realized that you probably meant all gov owners should be able to see all counters, not just the ones they own?

File: src/back-end/lib/resources/counter.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+type Resource = crud.Resource<
+  SupportedRequestBodies,
+  SupportedResponseBodies,
+  null,
+  null,
+  null,
+  null,
+  null,
+  null,
```
Comment (dhruvio) [inline src/back-end/lib/resources/counter.ts]: Need to use `UpdateRequestBody` here (I *think* it's this line)

File: src/back-end/lib/db/counter.ts (hunk 3 above)
Comment (dhruvio) [reply src/back-end/lib/db/counter.ts]: Yes please, the intention is just to increment counters

File: src/back-end/lib/permissions.ts (hunk 2 above)
Comment (dhruvio) [reply src/back-end/lib/permissions.ts]: Yeah, no need to do extra opp-based validation. Any gov user can view any counter.

**Comment by sutherlanda** [comment]
> One tiny change left, then we're good to merge.

Cool, updates done.

**Pull request merged**